service/src/main/java/org/cbioportal/service/impl/GenesetHierarchyServiceImpl.java

```diff
@@ -0,0 +1,385 @@
+			Integer percentile) {
+		
+		List<GenesetGeneticData> genesetScoreData = genesetScoresMap.get(geneset.getGenesetId());
+		List<GenesetGeneticData> genesetPvalueData = genesetPvaluesMap.get(geneset.getGenesetId());
+		
+		//return the maximum absolute value found:
+		List<ScoreAndPvalue> positiveScoresAndPvalues = new ArrayList<ScoreAndPvalue>();
+		List<ScoreAndPvalue> negativeScoresAndPvalues = new ArrayList<ScoreAndPvalue>();
+		
+		double max = 0;
+		double pvalueOfMax = 1;
+		for (int i = 0; i < genesetScoreData.size(); i++) {
+			String scoreString = genesetScoreData.get(i).getValue();
+			String pvalueString = genesetPvalueData.get(i).getValue();
+			
+			if (!NumberUtils.isNumber(scoreString)) 
+    			continue;
+    		
+			double score = Double.parseDouble(scoreString);
+			double pvalue = 1.0;
+			if (NumberUtils.isNumber(pvalueString))
+				pvalue = Double.parseDouble(pvalueString);
+			if (score >= 0) {
+				positiveScoresAndPvalues.add(new ScoreAndPvalue(score, pvalue));
+			} else {
+				negativeScoresAndPvalues.add(new ScoreAndPvalue(score, pvalue));
+			}
+			
+			//keep track of max, in case percentile is null
+    		if (Math.abs(score) > Math.abs(max)) {
+    			max = score; //here no abs, since we want to get the raw score (could be negative)
+    			pvalueOfMax = pvalue;
+    		}
+		}
+		
+		if (percentile == null) {
+			geneset.setRepresentativeScore(max);
+			geneset.setRepresentativePvalue(pvalueOfMax);
+		} else {
+			//sort scores:
+			Collections.sort(positiveScoresAndPvalues, ScoreAndPvalue.comparator);
+			Collections.sort(negativeScoresAndPvalues, ScoreAndPvalue.comparatorDesc);
+			
+			//use percentile:
+			ScoreAndPvalue representativePositiveScoreAndPvalue = new ScoreAndPvalue(0, 1);
+			ScoreAndPvalue representativeNegativeScoreAndPvalue = new ScoreAndPvalue(0, 1);
+			if (positiveScoresAndPvalues.size() > 0) {
+				int idxPositiveScores = (int)Math.round(percentile * positiveScoresAndPvalues.size() / 100.0);
+				representativePositiveScoreAndPvalue = positiveScoresAndPvalues.get(idxPositiveScores-1);
+			}
+			if (negativeScoresAndPvalues.size() > 0) {
+				int idxNegativeScores = (int)Math.round(percentile * negativeScoresAndPvalues.size() / 100.0);
+				representativeNegativeScoreAndPvalue = negativeScoresAndPvalues.get(idxNegativeScores-1);
+			}
+			
+			//set best one:
+			if (Math.abs(representativePositiveScoreAndPvalue.score) > Math.abs(representativeNegativeScoreAndPvalue.score)) {
+				geneset.setRepresentativeScore(representativePositiveScoreAndPvalue.score);
+				geneset.setRepresentativePvalue(representativePositiveScoreAndPvalue.pvalue);
+			} else {
+				geneset.setRepresentativeScore(representativeNegativeScoreAndPvalue.score);
+				geneset.setRepresentativePvalue(representativeNegativeScoreAndPvalue.pvalue);
+			}
+		}
+	}
+
+	/**
+	 * Inner class to hold score and p-value tuple. 
+	 * 
+	 * @author pieter
+	 *
+	 */
+	static class ScoreAndPvalue {
```

Comment:
I prefer not. I think it will be confusing. This is just a "transient" type to keep score and p-value together while we sort on score. At the end, the values are set in the geneset item: 
```java
geneset.setRepresentativeScore(representativeNegativeScoreAndPvalue.score);
geneset.setRepresentativePvalue(representativeNegativeScoreAndPvalue.pvalue);
```
Or do you know a better way that does not need this class perhaps?